snuba/datasets/entity.py

```diff
@@ -139,22 +137,64 @@ def build_match(
 
         if self._required_time_column:
             match = build_match(
-                self._required_time_column,
-                [
-                    ConditionFunctions.EQ,
-                    ConditionFunctions.GT,
-                    ConditionFunctions.GTE,
-                    ConditionFunctions.LT,
-                    ConditionFunctions.LTE,
-                ],
-                datetime,
+                self._required_time_column, [ConditionFunctions.EQ], datetime,
             )
             found = any(match.match(cond) for cond in top_level)
-            if not found:
+            if found:
+                return True
+
+            lower, upper = get_time_range_expressions(
+                top_level, self._required_time_column, alias
+            )
+            if not lower or not upper:
                 return False
 
+            # At this point we have valid conditions. However we need to align them and
+            # make sure they don't exceed the max_days. Replace the conditions.
+            self._replace_time_condition(query, *lower, *upper)
+
         return True
 
+    def _replace_time_condition(
+        self,
+        query: Query,
+        from_date: datetime,
+        from_exp: FunctionCall,
+        to_date: datetime,
+        to_exp: FunctionCall,
+    ) -> None:
+        max_days, date_align = state.get_configs(
+            [("max_days", None), ("date_align_seconds", 1)]
+        )
+
+        def align_fn(dt: datetime) -> datetime:
+            assert isinstance(date_align, int)
+            return dt - timedelta(seconds=(dt - dt.min).seconds % date_align)
+
+        from_date, to_date = align_fn(from_date), align_fn(to_date)
+        assert from_date <= to_date
+
+        if max_days is not None and (to_date - from_date).days > max_days:
+            from_date = to_date - timedelta(days=max_days)
+
+        def replace_cond(exp: Expression) -> Expression:
+            if not isinstance(exp, FunctionCall):
+                return exp
+            elif exp == from_exp:
+                return replace(
+                    exp, parameters=(from_exp.parameters[0], Literal(None, from_date)),
+                )
+            elif exp == to_exp:
+                return replace(
+                    exp, parameters=(to_exp.parameters[0], Literal(None, to_date))
+                )
```

Comment:
You may be transforming a time condition that was in an OR clause and that just happened to coincide with the one you are looking for. I assume you cannot just add the new conditions as the resulting SQL will be different from the original, So I think you could consider exploding the expression in top level conditions and only replacing the top level conditions that coincide with those you want to  replace? 
This instead of scanning the whole condition tree..

Reply:
Yeah that's true. I can change this to account for that.